### PR TITLE
Feature/red 10 tarball import

### PR DESCRIPTION
**Pull request by red-tt-bear**

### CODE REVIEW

File: src/features/courses/course-controller.ts [hunk 1]
```diff
@@ -1156,7 +1225,11 @@ class CourseController {
         let lastProblemNumber = await courseRepository.getLatestProblemNumberForTopic(options.courseTopicId) || 0;
         // TODO fix typings - remove any
         // eslint-disable-next-line @typescript-eslint/no-explicit-any
-        return useDatabaseTransaction<any>((): Promise<any> => {
+        return useDatabaseTransaction<any>(async (): Promise<any> => {
+            const userIds = options.userIds ?? await this.getEnrolledUserIdsInTopic({
```
Comment (doombot-exe) [inline src/features/courses/course-controller.ts:1229]: 🥇

File: src/features/courses/course-controller.ts [hunk 2]
```diff
@@ -4173,9 +4260,34 @@ You should be able to reply to the student's email address (${options.student.em
...
+                        // then check if it is already in the renderer
+                        if (ABSOLUTE_RENDERER_PATH_REGEX.test(pgFile.pgFilePathFromDefFile) === false && pgFile.pgFilePathFromDefFile.startsWith('local') === false) {
+                            const tryPath = `Contrib/${pgFile.pgFilePathFromDefFile}`;
+                            rendererAccessibleRequests++;
+                            const isAccessible = await rendererHelper.isPathAccessibleToRenderer({
+                                problemPath: tryPath
+                            });
+
+                            if (isAccessible) {
+                                // It's a secret to everyone #ZeldaReference
```
Comment (doombot-exe) [inline src/features/courses/course-controller.ts:4274]: :link-success:

**Review by doombot-exe — APPROVED** [review]

File: src/features/courses/course-controller.ts [hunk 3]
```diff
@@ -64,6 +64,8 @@ import { getAverageGroupsBeforeDate, QUESTION_SQL_NAME, STUDENTTOPICOVERRIDE_SQL
 // eslint-disable-next-line @typescript-eslint/no-var-requires
 const Sequelize = require('sequelize');
 
+const ABSOLUTE_RENDERER_PATH_REGEX = /^(:?private\/|Contrib\/|webwork-open-problem-library\/|Library\/)/;
```
Comment (PanopticaRising) [inline src/features/courses/course-controller.ts:67]: nit: You could probably pull the `\/` out to the end of the regex.

File: src/features/courses/course-controller.ts [hunk 4]
```diff
@@ -1148,6 +1150,73 @@ class CourseController {
...
+            }]
+        });
+
+        if (topics.length === 0) {
+            throw new RederlyError('getEnrolledUserIdsInTopic: fetch for topics had 0 results');
+        }
+        if (topics.length > 1) {
+            logger.warn('getEnrolledUserIdsInTopic fetched more than 1 topic, expected only one');
+        }
+        const unit = topics[0].unit;
```
Comment (PanopticaRising) [inline src/features/courses/course-controller.ts:1206]: Do we have `.first` on the backend?

File: src/features/courses/course-controller.ts [hunk 5]
```diff
@@ -4313,7 +4429,11 @@ You should be able to reply to the student's email address (${options.student.em
             }
             return unit;
         });
-        logger.info(`Import Course Archive complete ${new Date().getTime() - startTime} ${new Date()}`);
+        logger.info(`Import Course Archive complete ${JSON.stringify({
+            rendererSavePGFileRequests,
+            rendererSaveAssetRequests,
+            rendererAccessibleRequests,
+        })} ${new Date().getTime() - startTime} ${new Date()}`);
```
Comment (PanopticaRising) [inline src/features/courses/course-controller.ts:4436]: nit: Node should have `performance.now`. I don't know that it provides much accuracy benefit over the date object, though.

**Review by PanopticaRising — APPROVED** [review]

File: src/features/courses/course-controller.ts (hunk 3 above)
Comment (red-tt-bear) [reply src/features/courses/course-controller.ts:67]: Probably, I pulled it from @doombot-exe 's regex on the renderer

File: src/features/courses/course-controller.ts (hunk 4 above)
Comment (red-tt-bear) [reply src/features/courses/course-controller.ts:1206]: We do have .first, however, I was already checking length anyway so this is null safe (first would have made me check it again, I guess I could only check first in the 0 check earlier (probably would have but I refactored the code a bit and this isn't how it was initally organized)

File: src/features/courses/course-controller.ts (hunk 5 above)
Comment (red-tt-bear) [reply src/features/courses/course-controller.ts:4436]: Yeah agreed, eventually I plan to take this out anyway (or I guess I could demote to debug statements)

**Pull request merged**